src/headers/frame_header.rs

```diff
@@ -0,0 +1,494 @@
+#![allow(clippy::excessive_precision)]
+
+use crate::{bit_reader::BitReader, error::Error, headers::encodings::*};
+
+use jxl_headers_derive::UnconditionalCoder;
+use num_derive::FromPrimitive;
+
+#[derive(UnconditionalCoder, Copy, Clone, PartialEq, Debug, FromPrimitive)]
+enum FrameType {
+    RegularFrame = 0,
+    LFFrame = 1,
+    ReferenceOnly = 2,
+    SkipProgressive = 3,
+}
+
+#[derive(UnconditionalCoder, Copy, Clone, PartialEq, Debug, FromPrimitive)]
+enum Encoding {
+    VarDCT = 0,
+    Modular = 1,
+}
+
+#[derive(UnconditionalCoder, Debug, PartialEq)]
+struct Passes {
+    #[coder(u2S(1, 2, 3, Bits(3) + 4))]
+    #[default(1)]
+    num_passes: u32,
+
+    #[coder(u2S(0, 1, 2, Bits(1) + 3))]
+    #[default(0)]
+    #[condition(num_passes != 1)]
+    num_ds: u32,
+
+    #[size_coder(explicit(num_passes - 1))]
+    #[coder(Bits(2))]
+    #[default(Vec::new())]
+    #[condition(num_passes != 1)]
+    shift: Vec<u32>,
+
+    #[size_coder(explicit(num_ds))]
+    #[coder(u2S(1, 2, 4, 8))]
+    #[default(Vec::new())]
+    #[condition(num_passes != 1)]
+    downsample: Vec<u32>,
+
+    #[size_coder(explicit(num_ds))]
+    #[coder(u2S(0, 1, 2, Bits(3)))]
+    #[default(Vec::new())]
+    #[condition(num_passes != 1)]
+    last_pass: Vec<u32>,
+}
+
+#[derive(UnconditionalCoder, Copy, Clone, PartialEq, Debug, FromPrimitive)]
+enum BlendingMode {
+    Replace = 0,
+    Add = 1,
+    Blend = 2,
+    AlphaWeightedAdd = 3,
+    Mul = 4,
+}
+
+struct BlendingInfoNonserialized {
+    num_extra_channels: u32,
+    have_crop: bool,
+    x0: u32,
+    y0: u32,
+    width: u32,
+    height: u32,
+    img_width: u32,
+    img_height: u32,
+}
+
+#[derive(UnconditionalCoder, Debug, PartialEq, Clone)]
+#[nonserialized(BlendingInfoNonserialized)]
+struct BlendingInfo {
+    #[coder(u2S(0, 1, 2, Bits(2) + 3))]
+    #[default(BlendingMode::Replace)]
+    mode: BlendingMode,
+
+    /* Spec: "Let multi_extra be true if and only if and the number of extra channels is at least two."
+    libjxl condition is num_extra_channels > 0 */
+    #[coder(u2S(0, 1, 2, Bits(3) + 3))]
+    #[default(0)]
+    #[condition(nonserialized.num_extra_channels > 0 &&
+        (mode == BlendingMode::Blend || mode == BlendingMode::AlphaWeightedAdd))]
+    alpha_channel: u32,
+
+    #[default(false)]
+    #[condition(nonserialized.num_extra_channels > 0 &&
+        (mode == BlendingMode::Blend || mode == BlendingMode::AlphaWeightedAdd || mode == BlendingMode::Mul))]
+    clamp: bool,
+
+    #[coder(u2S(0, 1, 2, 3))]
+    #[default(0)]
+    /* Let full_frame be true if and only if have_crop is false or if the
+    frame area given by width and height and offsets x0 and y0 completely covers the image area. */
```

Comment:
Wishlist for the future: being

Reply:
It bothers me too, especially when part of the condition such as `full_frame` has to be duplicated. 
We could introduce a new attribute on top of the struct like `#[free_variables(foo = bar, ...)]`. But I'm just going to add a TODO there for now.

Reply:
Yeah, although I'm not sure how it would work at the top of the struct - I was thinking of adding new members with a #[computed(...)] attribute, which should work without too many problems but would likely waste some amount of space and potentially create inconsistent states...